File: python/monocr_onnx/segmenter.py
import cv2
import numpy as np
from PIL import Image

class LineSegmenter:
    """
    Robust line segmenter using Horizontal Projection Profiles with Smoothing.
    Handles noisy documents and touching lines by finding valleys in the projection.
    """
    def __init__(self, smooth_kernel=5, threshold_ratio=0.02):
        self.smooth_kernel = smooth_kernel
        self.threshold_ratio = threshold_ratio

    def segment(self, image):
        """
        Segment a document image into text lines.
        Args:
            image (PIL.Image or np.ndarray): Input image.
        Returns:
            list: List of dicts with keys 'img' (PIL.Image) and 'bbox' (x, y, w, h).
        """
        if isinstance(image, Image.Image):
            img_pil = image
            if img_pil.mode != 'L':
                img_pil = img_pil.convert('L')
            img_arr = np.array(img_pil)
        else:
            img_arr = image
            img_pil = Image.fromarray(img_arr)
            if img_pil.mode != 'L':
                img_pil = img_pil.convert('L')
                img_arr = np.array(img_pil)

        # 1. Binarize (Adaptive Thresholding)
        # Invert so text is white, background black
        binary = cv2.adaptiveThreshold(
            img_arr, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, 
            cv2.THRESH_BINARY_INV, 25, 10
        )
        
        # 2. Horizontal Projection Profile
        hist = np.sum(binary, axis=1) # Sum along width (axis 1) -> shape (height,)

        # 3. Smoothing
        if self.smooth_kernel > 1:
            kernel = np.ones(self.smooth_kernel) / self.smooth_kernel
            # mode='same' returns output of same length as input
            smoothed_hist = np.convolve(hist, kernel, mode='same')
        else:
            smoothed_hist = hist

        # 4. Gap Detection
        non_zero_vals = smoothed_hist[smoothed_hist > 0]
        if len(non_zero_vals) == 0:
            return []

        # Find max density 
        max_val = np.max(smoothed_hist)
        threshold = max_val * self.threshold_ratio # e.g. 2% of max density is a gap
        
        results = []
        start = None
        height, width = img_arr.shape
        
        for y in range(height):
            is_text = smoothed_hist[y] > threshold
            
            if is_text and start is None:
                start = y
            elif not is_text and start is not None:
                end = y
                if (end - start) > 8: # Minimal line height check
                    # Add generous padding
                    pad = 4
                    y1 = max(0, start - pad)
                    y2 = min(height, end + pad)
                    
                    # Crop full width
                    crop = img_pil.crop((0, y1, width, y2))
                    
                    results.append({
                        'img': crop,
                        'bbox': (0, int(y1), int(width), int(y2 - y1))
                    })
                start = None
                
        # Handle last segment
        if start is not None and (height - start) > 8:
            pad = 4
            y1 = max(0, start - pad)
            y2 = min(height, height)
            crop = img_pil.crop((0, y1, width, y2))
            
            results.append({
                'img': crop,
                'bbox': (0, int(y1), int(width), int(y2 - y1))
            })
            
        return results

File: go/pkg/segmenter/segmenter.go
package segmenter

import (
	"image"
	"image/color"
	"image/draw"
	"math"
)

type LineSegmenter struct {
	MinLineH     int
	SmoothWindow int
}

type SegmentResult struct {
	Img  image.Image
	BBox image.Rectangle
}

func NewLineSegmenter(minLineH, smoothWindow int) *LineSegmenter {
	if minLineH == 0 {
		minLineH = 10
	}
	if smoothWindow == 0 {
		smoothWindow = 3
	}
	return &LineSegmenter{
		MinLineH:     minLineH,
		SmoothWindow: smoothWindow,
	}
}

func (s *LineSegmenter) Segment(img image.Image) ([]SegmentResult, error) {
	// Convert to Grayscale if needed (conceptually, we just need luminance)
	bounds := img.Bounds()
	width, height := bounds.Dx(), bounds.Dy()

	// 1. Horizontal Projection Profile
	// We want to count 'text' pixels (dark pixels < 128)
	// hist[y] = sum(is_text(x, y) for x in width)
	hist := make([]int, height)

	// Accessing pixels via At() is slow, but compatible with all image types.
	// For optimization later, type switch to *image.Gray, *image.RGBA etc.
	// For now, simple implementation.
	for y := 0; y < height; y++ {
		sum := 0
		for x := 0; x < width; x++ {
			c := img.At(bounds.Min.X+x, bounds.Min.Y+y)
			gray := color.GrayModel.Convert(c).(color.Gray)
			if gray.Y < 128 {
				sum++
			}
		}
		hist[y] = sum
	}

	// 2. Smoothing
	smoothedHist := make([]float64, height)
	if s.SmoothWindow > 1 {
		window := float64(s.SmoothWindow)
		overflow := s.SmoothWindow / 2
		for i := 0; i < height; i++ {
			sum := 0.0
			count := 0.0
			for k := -overflow; k <= overflow; k++ {
				idx := i + k
				if idx >= 0 && idx < height {
					sum += float64(hist[idx])
					count++
				}
			}
			smoothedHist[i] = sum / window // Python code divides by window size (kernel is 1/window)
			// Actually numpy convolution handling at edges: 'same' mode zero-pads?
			// Python: np.convolve(hist, np.ones(window)/window, mode='same')
			// Let's stick to a simple moving average.
		}
	} else {
		for i, v := range hist {
			smoothedHist[i] = float64(v)
		}
	}

	// 3. Gap Detection
	var nonZeroVals []float64
	for _, v := range smoothedHist {
		if v > 0 {
			nonZeroVals = append(nonZeroVals, v)
		}
	}

	if len(nonZeroVals) == 0 {
		return []SegmentResult{}, nil
	}

	// Mean density
	sumVal := 0.0
	for _, v := range nonZeroVals {
		sumVal += v
	}
	meanDensity := sumVal / float64(len(nonZeroVals))
	gapThreshold := meanDensity * 0.05

	var results []SegmentResult
	var start *int

	// 4. Extract Lines
	for y := 0; y < height; y++ {
		isText := smoothedHist[y] > gapThreshold

		if isText && start == nil {
			s := y
			start = &s
		} else if !isText && start != nil {
			end := y
			if (end - *start) >= s.MinLineH {
				s.extractLine(img, bounds, *start, end, &results)
			}
			start = nil
		}
	}

	if start != nil && (height-*start) >= s.MinLineH {
		s.extractLine(img, bounds, *start, height, &results)
	}

	return results, nil
}

func (s *LineSegmenter) extractLine(img image.Image, bounds image.Rectangle, rStart, rEnd int, results *[]SegmentResult) {
	// Find horizontal bounds within strip
	// strip corresponds to y inside [bounds.Min.Y + rStart, bounds.Min.Y + rEnd)
	// We need to sum columns to find x range.

	width := bounds.Dx()
	colSum := make([]int, width)

	// Optimize: Only loop through the strip rows
	for y := rStart; y < rEnd; y++ {
		actualY := bounds.Min.Y + y
		for x := 0; x < width; x++ {
			actualX := bounds.Min.X + x
			c := img.At(actualX, actualY)
			gray := color.GrayModel.Convert(c).(color.Gray)
			if gray.Y < 128 {
				colSum[x]++
			}
		}
	}

	// Find non-empty cols
	xMin := -1
	xMax := -1

	for x := 0; x < width; x++ {
		if colSum[x] > 0 {
			if xMin == -1 {
				xMin = x
			}
			xMax = x
		}
	}

	if xMin == -1 {
		return
	}

	// Add padding
	pad := 4
	y1 := int(math.Max(0, float64(rStart-pad)))
	y2 := int(math.Min(float64(bounds.Dy()), float64(rEnd+pad)))
	x1 := int(math.Max(0, float64(xMin-pad)))
	x2 := int(math.Min(float64(width), float64(xMax+pad)))

	// Crop
	rect := image.Rect(bounds.Min.X+x1, bounds.Min.Y+y1, bounds.Min.X+x2, bounds.Min.Y+y2)

	// Since Go images merely reference the underlying buffer when sub-imaging,
	// and we might want to resize/process these individually,
	// we should probably copy to a new buffer or just return the SubImage.
	// SubImage is safer for memory if original is large, but for processing we might want clean buffer.
	// Let's return a SubImage if possible, or a Copy. LineSegmenter usually returns independent images
	// in Python PIL (crop returns copy).
	// onnx runtime preprocessing will maximize contrast etc anyway?
	// Actually `image.NewGray` and draw is safest to ensure 'L' mode equivalent.

	dst := image.NewGray(image.Rect(0, 0, x2-x1, y2-y1))
	draw.Draw(dst, dst.Bounds(), img, rect.Min, draw.Src)

	*results = append(*results, SegmentResult{
		Img:  dst,
		BBox: rect,
	})
}
